### PR TITLE
Redirect printf to USB logging and support USB TASH

### PR DESCRIPTION
**Pull request by zhongnuo-tang**

### CODE REVIEW

File: os/board/rtl8730e/src/Makefile [hunk 1]
```diff
@@ -120,6 +120,12 @@ include component/bluetooth/example/ble_scatternet/Make.defs
 endif
```
Comment (edwakuwaku) [inline os/board/rtl8730e/src/Makefile:120]: Please change file permission back to 644

File: os/board/rtl8730e/src/rtl8730e_boot.c [hunk 2]
```diff
@@ -481,17 +485,8 @@ void board_initialize(void)
 #endif
 	/* Enable IPC buffered print */
 	inic_ipc_buffered_printf_set_np_enable(1);
-
 #ifdef CONFIG_AUDIO_ALC1019
-	if (rtl8730e_alc1019_initialize(0) != 0) {
```
Comment (edwakuwaku) [inline os/board/rtl8730e/src/rtl8730e_boot.c:486]: Please revert this

File: os/arch/arm/src/amebasmart/amebasmart_usb.c [hunk 3]
```diff
@@ -0,0 +1,803 @@
...
+{
+	struct rtl8730e_up_dev_s *priv = (struct rtl8730e_up_dev_s *)dev->priv;
+	DEBUGASSERT(priv);
+	uint16_t nbytes = 0;
+	uint8_t *buffer = NULL;
+	buffer = (uint8_t *)rtw_zmalloc(CONFIG_CDC_ACM_BULK_IN_XFER_SIZE);
+	if (buffer == NULL) {
+		dbg("malloc failed\n");
+	}
+#ifdef CONFIG_SMP
```
Comment (edwakuwaku) [inline os/arch/arm/src/amebasmart/amebasmart_usb.c:342]: No need to mark macro, in single core case, enter/leave_critical_section will be redirected to irqsave/restore

File: os/arch/arm/src/amebasmart/amebasmart_usb.c [hunk 4]
```diff
@@ -0,0 +1,803 @@
...
+	}
+	USB_DEV.isconsole = true;
+	uart_register("/dev/ttyACM0", &USB_DEV);
+	unregister_driver("/dev/console");
+	uart_register("/dev/console", &USB_DEV);
+	int fd;
+	fd = open("/dev/console", O_RDWR);
+	if(fd >= 0) {
+		dup2(fd, 1);
+    	dup2(fd, 2);
```
Comment (edwakuwaku) [inline os/arch/arm/src/amebasmart/amebasmart_usb.c:778]: alignment

File: os/arch/arm/src/amebasmart/amebasmart_usb.c [hunk 5]
```diff
@@ -0,0 +1,803 @@
...
+ register with serial first to prevent opening an empty path, then once usb is ready, re-register it to usb and
+ open the fd*/
+static int register_usb(void)
+{
+	while(!cdc_acm_ready_flag){
+		usleep(1000);
+	}
+	USB_DEV.isconsole = true;
+	uart_register("/dev/ttyACM0", &USB_DEV);
+	unregister_driver("/dev/console");
```
Comment (edwakuwaku) [inline os/arch/arm/src/amebasmart/amebasmart_usb.c:772]: Let's add some comment on why we are unregistering this, eg; due to serial sharing same fd........

**Comment by edwakuwaku** [comment]
Hello @r-prabu  @sunghan-chang , this PR is tested with rerouting printf with USB. Let's verify and check together

**Comment by r-prabu** [comment]
> Hello @r-prabu @sunghan-chang , this PR is tested with rerouting printf with USB. Let's verify and check together

> ok, will verify from our side and update the results.

**Comment by r-prabu** [comment]
> > Hello @r-prabu @sunghan-chang , this PR is tested with rerouting printf with USB. Let's verify and check together
> 
> > ok, will verify from our side and update the results.

@edwakuwaku during the initial booting below log messages are routed to UART & rest of the logs are coming on USB. Is there any way to fix the same to print the initial logs also over USB?

Normal boot
flash_size: 16M
BP1 data valid, version:1
BP2 version invalid
BP1 CRC32 match, attached CRC: 0x4698b6a1, calculated CRC: 0x4698b6a1
BP1 selected
Active Index: 0
OTA1 CRC32 match, attached CRC: 0xcfa9c7b3, calculated CRC: 0xcfa9c7b3
OTA1 USE, version: 200204
[BOOT-I] KM0 XIP IMG[0c000000:20]
[BOOT-I] KM0 SRAM[23002000:114a0]
[BOOT-I] KM0 DRAM[6fffffdf:20]
[BOOT-I] KM4 XIP IMG[0d000000:535c0]
[BOOT-I] KM4 SRAM[20014000:40]
[BOOT-I] KM4 DRAM[60000000:800]
[MODULE_BOOT-LEVEL_INFO]:IMG2(OTA1) VALID, ret: 0
[BOOT-I] AP XIP IMG[0e000000:c33c0]
[BOOT-I] AP BL1 SRAM[3001fde0:40]
[BOOT-I] AP BL1 DRAM[70017000:3709]
[BOOT-I] AP FIP[601fffe0:43315]
[BOOT-I] RDP OFF
[BOOT-I] AP FIP[601fffe0:43315]
[APP-I] KM0 APP_START
[APP-I] KM0 CPU CLK: 40000000 Hz
[APP-I] KM0 VTOR:0x23000000
[PMC-I] NP wake event: 2001000 0
[PMC-I] AP wake event 8400c080 40
[PMC-I] LP wake event 80001b 20
[MAIN-I] KM0 OS START
up_allocate_kheap: start = 0x6011a700 size = 7215360
up_initialize: [Reboot Reason] : 55
os_start: All SMP cores started!!
[USB-I] USB init
[USB-I] USB configurations:
[USB-I] * speed: 3
[USB-I] * dma_enable: 1
[USB-I] * isr_priority: 25
[USB-I] * intr_use_ptx_fifo: 0
[USB-I] * ptx_fifo_first: 0
[USB-I] * ext_intr_en: 0x00000004
[USB-I] * nptx_max_epmis_cnt: 10
[USB-I] * nptx_max_err_cnt[0]: 0
[USB-I] * nptx_max_err_cnt[1]: 0
[USB-I] * nptx_max_err_cnt[2]: 0
[USB-I] * nptx_max_err_cnt[3]: 2000
[USB-I] * nptx_max_err_cnt[4]: 0
[USB-I] * nptx_max_err_cnt[5]: 0
[USB-I] GSNPSID = 0x4F54310A
uart_register: Registering /dev/ttyACM0
uart_register: Registering /dev/console
up_flashinitialize: Manufacturer : 239 memory type : 64 capacity : 24
amebasmart_cdc_acm_cb_setup:
[CDC] VCOM port activated
[rtl_readdecrypt_factorykey] Exit form here!
Initializing WIFI ...[OTP-I] OTP_LogicalMap_Read: data end at address=88
[RTW]: ** Ba
nd = 2.4G and 5G **
KM4 version:  km4_application_ver_4eade3d8b7_2024/08/27-14:57:56

**Comment by kishore-sn** [comment]
I notice two issues when testing this PR:
1. When i press the reset button on the board, then the ttyACM device gets disconnected and reconnected. But with uart port, this problem is not there. If we are using some tool like putty then we will not get the initial boot logs because putty session gets closed when the port gets disconnected.

2. When I check with linux minicom tool, I am seeing mixed logs during boot up. Is this due to some wrong setting like speed? What baud rate are we supposed to use with this?

File: os/arch/arm/src/amebasmart/amebasmart_usb.c [hunk 6]
```diff
@@ -0,0 +1,803 @@
+/****************************************************************************
```
Comment (edwakuwaku) [inline os/arch/arm/src/amebasmart/amebasmart_usb.c:1]: Many of the variable/function requires cleanup and more comment explanation. But due to tight schedule, let's do it in a new PR.

File: os/arch/arm/src/amebasmart/amebasmart_usb.c [hunk 7]
```diff
@@ -0,0 +1,803 @@
...
+	struct rtl8730e_up_dev_s *priv = (struct rtl8730e_up_dev_s *)dev->priv;
+	DEBUGASSERT(priv);
+	uint16_t nbytes = 0;
+	uint8_t *buffer = NULL;
+	buffer = (uint8_t *)rtw_zmalloc(CONFIG_CDC_ACM_BULK_IN_XFER_SIZE);
+	if (buffer == NULL) {
+		lldbg("malloc failed\n");
+	}
+
+	irqstate_t flags = enter_critical_section();
```
Comment (sunghan-chang) [inline os/arch/arm/src/amebasmart/amebasmart_usb.c:343]: leave_critical_section is in the CONFIG_SMP conditional. But here is no. Need to be in the same conditional.

File: os/arch/arm/src/amebasmart/amebasmart_usb.c [hunk 8]
```diff
@@ -0,0 +1,803 @@
...
+	} else {
+		printf("\n[CDC] Busy, discarded %d bytes\n", len);
+		ret = HAL_BUSY;
+	}
+
+	return ret;
+#else
+	receive_data = (uint8_t)*buf;
+	received_flag = 1;
+	uart_recvchars(&USB_DEV);
```
Comment (sunghan-chang) [inline os/arch/arm/src/amebasmart/amebasmart_usb.c:599]: uart??

File: os/board/rtl8730e/src/component/usb/device_new/core/usbd_pcd.c [hunk 9]
```diff
@@ -0,0 +1,2029 @@
...
+			return ret;
+		}
+	}
+
+	pcd->pcd_state = HAL_PCD_STATE_BUSY;
+	pcd->phy_initialized = 0;
+	pcd->isr_initialized = 0;
+	pcd->setup = (u8 *)usb_os_mem_alloc(USBD_SETUP_PACKET_BUF_LEN);
+	if (pcd->setup == NULL) {
+		return HAL_ERR_MEM;
```
Comment (sunghan-chang) [inline os/board/rtl8730e/src/component/usb/device_new/core/usbd_pcd.c:1515]: Don't you need to free allocated memory before? I mean `pcd->lock`.
same as other returns

**Comment by Taejun-Kwon** [comment]
It should work with usb cdc/acm driver in below.
https://github.com/Samsung/TizenRT/tree/master/os/drivers/usbdev

**Comment by edwakuwaku** [comment]
> It should work with usb cdc/acm driver in below. https://github.com/Samsung/TizenRT/tree/master/os/drivers/usbdev

Hello @Taejun-Kwon , this has been discussed via email before. Due to below reasons,
1) Tight schedule, insufficient time to integrate with TizenLite cdcacm dev driver
2) TizenLite USB device cdc acm class has not been verified before

This PR is to route USB through /dev/console to replace UART instead of /dev/ttyACM. We seek for your understanding and please accept this current version as to tackle the performance bottleneck due to majority logs from application side.

File: os/arch/arm/src/amebasmart/amebasmart_usb.c (hunk 8 above)
Comment (sunghan-chang) [reply os/arch/arm/src/amebasmart/amebasmart_usb.c:599]: Let's clean this up later by using usb cdc/acm driver.

File: os/arch/arm/src/amebasmart/amebasmart_usb.c (hunk 3 above)
Comment (zhongnuo-tang) [reply os/arch/arm/src/amebasmart/amebasmart_usb.c]: Hi @sunghan-chang,
Should we add back the macro or change to irqsave/restore?

**Review by ewoodev — COMMENTED** [review]
The USB console should also be registered as a wakeup source about the power management
Please let me know if you would like to update this pull request or create a new one.

And this is a later story, but when we use multiple devices with one USB port such as in a USB hub, can we wake them up separately?

File: os/board/rtl8730e/src/component/usb/device_new/core/usbd_pcd.c (hunk 9 above)
Comment (zhongnuo-tang) [reply os/board/rtl8730e/src/component/usb/device_new/core/usbd_pcd.c]: Free allocated memory when error occurred

File: os/arch/arm/src/amebasmart/amebasmart_usb.c (hunk 7 above)
Comment (zhongnuo-tang) [reply os/arch/arm/src/amebasmart/amebasmart_usb.c:343]: added CONFIG_SMP

**Comment by zhongnuo-tang** [comment]
> The USB console should also be registered as a wakeup source about the power management Please let me know if you would like to update this pull request or create a new one.
--- > this might need some time to check on the implementation as well as measurement of the power usage, I think it would be better to have another PR for USB power management. 
> And this is a later story, but when we use multiple devices with one USB port such as in a USB hub, can we wake them up separately?
--- > i will check on this.

**Review by ewoodev — APPROVED** [review]

**Review by sunghan-chang — APPROVED** [review]

**Pull request merged**